libsolidity/formal/CHC.cpp

```diff
@@ -83,16 +85,19 @@ void CHC::analyze(SourceUnit const& _source)
 	checkVerificationTargets();
 
 	bool ranSolver = true;
-#ifndef HAVE_Z3
-	ranSolver = dynamic_cast<CHCSmtLib2Interface const*>(m_interface.get())->unhandledQueries().empty();
-#endif
+	if (auto const* smtLibInterface = dynamic_cast<CHCSmtLib2Interface const*>(m_interface.get()))
+		ranSolver = smtLibInterface->unhandledQueries().empty();
 	if (!ranSolver && !m_noSolverWarning)
 	{
 		m_noSolverWarning = true;
 		m_outerErrorReporter.warning(
 			3996_error,
 			SourceLocation(),
+#ifdef HAVE_Z3_DLOPEN
+			"CHC analysis was not possible since no suitable version of libz3.so was found."
+#else
 			"CHC analysis was not possible since no integrated z3 SMT solver was found."
```

Comment:
Cannot this be a single version: "CHC analysis was not possible since no z3 SMT solver was found." ?

Is there a benefit to have the ability tell what kind of z3 is missing? If so, we should test for it. If not, we should change to a uniform message.

Reply:
For sure there is a benefit in knowing what kind is missing - how else would you know what kind you need to install?
But that's *only* an issue for dynamic loading - if there is no dynamic loading, you can't just install anything anyways, so you're out of luck anyways - in that case there *is* no version to report and even if there were it wouldn't help.

Not sure what to test in any of that.

Reply:
Have a test case where dynamic loading is enabled but no (appropriate) libz3.so is found?

Reply:
Somehow missed that last comment - but yeah, that's not so much a test, but an entire test run then and we have an issue for that in https://github.com/ethereum/solidity/issues/10519 - we definitely need such a test run and in that context can cover this here as well.